### PR TITLE
[dev] Global hotkey refactor

### LINKED ISSUE #788: Win+R with Shell plugin doesn't work well (with windows walker)
**Describe the bug/issue**
- Window close when press win+r after 1.83 ver (dev)

**To Reproduce**
1. Open Flow launcher 
2. Press Win+R
3. Check that the window is hidden.

Occurred in Dev version. Normal operation at 1.83. One of the modifications appears to have an effect.

### PR DESCRIPTION
**Pull request by taooceros**
fix #788

### CODE REVIEW

**Comment by jjw24** [comment]
@taooceros fair bit of method signature changes. can you give me a quick overview of why they are changed please

**Comment by taooceros** [comment]
> @taooceros fair bit of method signature changes. can you give me a quick overview of why they are changed please

Only the global hotkey event in publicapi is obsolete, and two new api is added to handle the same work.

**Comment by taooceros** [comment]
Unsafe is used because
1. It's reasonable to have high performance here because every key hit will trigger the running of the function once.
2. I would like to try the new function pointer feature🤔😂
3. I am not sure whether some issues may exist because of unsafe, so we should take a small set of testing here. However, it is unlikely we need to unload assembly for infrastructure, so it should be fine.

**Review by onesounds — DISMISSED** [review]

**Review by Garulf — APPROVED** [review]

File: Plugins/Flow.Launcher.Plugin.Shell/Main.cs [hunk 1]
```diff
@@ -337,12 +337,9 @@ bool API_GlobalKeyboardEvent(int keyevent, int vkcode, SpecialKeyState state)
 
         private void OnWinRPressed()
         {
-            context.API.ChangeQuery($"{context.CurrentPluginMetadata.ActionKeywords[0]}{Plugin.Query.TermSeparator}");
-
             // show the main window and set focus to the query box
-            Window mainWindow = Application.Current.MainWindow;
-            mainWindow.Show();
-            mainWindow.Focus();
```
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Shell/Main.cs:345]: does this still need focus applied?

File: Flow.Launcher.Infrastructure/Hotkey/GlobalHotkey.cs [hunk 2]
```diff
@@ -91,17 +78,17 @@ private IntPtr LowLevelKeyboardProc(int nCode, UIntPtr wParam, IntPtr lParam)
             {
                 return InterceptKeys.CallNextHookEx(hookId, nCode, wParam, lParam);
             }
-            return (IntPtr)1;
+            return (IntPtr)(-1);
```
Comment (jjw24) [inline Flow.Launcher.Infrastructure/Hotkey/GlobalHotkey.cs:81]: what's this change?

**Comment by jjw24** [comment]
what's been tested with this change so far?

also need to bump Shell plugin

**Comment by onesounds** [comment]
> what's been tested with this change so far?
> 
> also need to bump Shell plugin

It is a pr to solve the problem that when win+r is pressed, the flow closes and the run command (win+r by os) operates. I understand that the code has become complicated because there is a part in the middle of the work that does not work properly related to animation.

**Comment by jjw24** [comment]
well i dont exactly understand this change, so if @taooceros is sure this is good to go and not going to break anything then happy for one of you to approve and merge this.

**Comment by taooceros** [comment]
> well i dont exactly understand this change, so if @taooceros is sure this is good to go and not going to break anything then happy for one of you to approve and merge this.

The main change is because the multicast delegate (event) fails to return all return values so that the win+r may not work with the presence of the windows walker. The main change of the pull request change the event to a list of Func to retrieve the return value correctly.
Also, because c# 9 added the function pointer, so I think it's cool to use that in the hook for better performance because the hook is fired in every single keyboard click.

**Review by jjw24 — CHANGES_REQUESTED** [review]
somehow the artifacts are smaller, doesnt seem normal.

image

**Comment by Garulf** [comment]
- Build looks good
- <kbd>Win</kbd>+<kbd>R,/kbd> fixed
- Fix for Tab complete
- Working with Window Walker
- No crashes

Good to go?

**Review by jjw24 — APPROVED** [review]
unsure of code changes but tested working fine, will merge since this addresses an issue in current dev

**Pull request merged**